src/StaticATokenLM.sol

```diff
@@ -415,6 +416,64 @@ contract StaticATokenLM is
     return balanceOf[owner];
   }
 
+  function maxRedeemUnderlying(address owner)
+    public
+    view
+    virtual
+    returns (uint256)
+  {
+    address cachedATokenUnderlying = _aTokenUnderlying;
+    DataTypes.ReserveData memory reserveData = POOL.getReserveData(
+      cachedATokenUnderlying
+    );
+
+    // if paused or inactive users cannot withraw underlying
+    if (
+      !ReserveConfiguration.getActive(reserveData.configuration) ||
+      ReserveConfiguration.getPaused(reserveData.configuration)
+    ) {
+      return 0;
+    }
+
+    // otherwise users can withdraw up to the available amount
+    uint256 aTokenBalance = IERC20(cachedATokenUnderlying).balanceOf(
+      reserveData.aTokenAddress
+    );
+    uint256 cachedUserBalance = balanceOf[owner];
+    return
+      aTokenBalance >= _convertToAssets(cachedUserBalance, Rounding.DOWN)
+        ? cachedUserBalance
+        : _convertToShares(aTokenBalance, Rounding.DOWN);
+  }
+
+  // reference: https://github.com/aave/aave-v3-core/blob/a00f28e3ad7c0e4a369d8e06e0ac9fd0acabcab7/contracts/protocol/libraries/logic/ValidationLogic.sol#L57
+  function maxDepositUnderlying() public view virtual returns (uint256) {
+    DataTypes.ReserveData memory reserveData = POOL.getReserveData(
+      _aTokenUnderlying
+    );
+
+    // if inactive, paused or frozen users cannot deposit underlying
+    if (
+      !ReserveConfiguration.getActive(reserveData.configuration) ||
+      ReserveConfiguration.getPaused(reserveData.configuration) ||
+      ReserveConfiguration.getFrozen(reserveData.configuration)
+    ) {
+      return 0;
+    }
+
+    uint256 supplyCap = ReserveConfiguration.getSupplyCap(
+      reserveData.configuration
+    ) * (10**ReserveConfiguration.getDecimals(reserveData.configuration));
+    // if no supply cap deposit is unlimited
+    if (supplyCap == 0) return type(uint256).max;
+    // return remaining supply cap margin
+    uint256 currentSupply = (IAToken(reserveData.aTokenAddress)
+      .scaledTotalSupply() + reserveData.accruedToTreasury).rayMulRoundUp(
```

Comment:
this implementation is not 100% equal as `ValidationLogic` uses the halfup rounding as always.
Shouldn't matter in that case though as it's rounding on a wei for a result that has 0 wei precision (the caps are extended with 18 decimals in that case.

Reply:
I think that in the case of a deposit is good to use round-up as you did. We don't mind that the user will be allowed to deposit 1 wei less than it's allowed by the system